### PR TITLE
suppress alias warnings with verbosity<0 (fixes #4518)

### LINKED ISSUE #4518: Unable to suppress parameter alias related warnings
## Description

Many parameter aliases trigger warnings that may be annoying and repetitive, specially in the context of model selection using grid search. These warnings can't be silenced except by changing the parameter to some specific alias. Often, this specific alias is not the one "recommended" (most used or first listed) in the documentation.

## Reproducible example

See https://github.com/microsoft/LightGBM/issues/3641#issuecomment-898092405

## Environment info

LightGBM version or commit hash: 3.2.1

### PR DESCRIPTION
**Pull request by jmoralez**
Sets the logging level before resolving the parameter aliases so that the warnings related to them can be suppressed using the verbosity parameter.

Fixes #4518.

### CODE REVIEW

File: tests/python_package_test/test_engine.py
Comment (jameslamb) [inline tests/python_package_test/test_engine.py:3583]: Thanks for this test!

I'm ok with hard-coding `"verbose"` followed by `"verbosity"` in `config.cpp`, as I think that's more efficient than calling  `LGBM_DumpParamAliases()` (introduced in #4829) or something like it to get all parameter aliases that control verbosity.

But could you modify this test such that it would catch the case "LightGBM doesn't respect all aliases for managing verbosity"? That way, we'd know to modify `config.cpp` if 

I think that could be accomplished by changing the test structure to something like the following:

```python
@pytest.mark.parametrize('verbosity_param', _ConfigAliases.get("verbose"))
@pytest.mark.parametrize('verbosity_value', [-1, 0])
def test_verbosity_can_suppress_alias_warnings(capsys, verbosity_param, verbosity_value):
    # ...
    params = {
        # ...
        verbosity_param: verbosity_value
        # ...
    }
    # ... 
```

File: src/io/config.cpp
Comment (jameslamb) [inline src/io/config.cpp:57]: Super minor thing, but would you consider re-ordering this such that it is order by verbosity (either from most to least verbose or vice-versa)? Right now it's jumping back and forth.

e.g., I think it's be slightly easier to understand if it went `DEBUG --> INFO --> WARN --> FATAL` or `FATAL --> WARN --> INFO --> DEBUG`

File: tests/python_package_test/test_engine.py
Comment (jameslamb) [inline tests/python_package_test/test_engine.py:3601]: Instead of testing that `stdout` is empty, I think this should be changed to checking that there are 0 LightGBM-generated messages (i.e. those containing `[LightGBM]`) emitted.

That would prevent the case where this test fails when one of `lightgbm`'s dependencies emits some log messages (like a deprecation warning or something) that `lightgbm` does not try to suppress based on the value of its `verbosity` setting.

I proposed something similar in the R package's tests when working on #4862, e.g. https://github.com/microsoft/LightGBM/blob/fb37e5074ad0163afd9d88366b58075899775cfa/R-package/tests/testthat/test_basic.R#L2312-L2315

**Review by jameslamb — CHANGES_REQUESTED** [review]
Thanks for doing this! Like I mentioned in https://github.com/microsoft/LightGBM/issues/4518#issuecomment-1140363388, I totally support this change. But I'd like to hear @StrikerRUS's perspective on this one as well before this is merged.

Until them, please see a few small suggestions below.

File: src/io/config.cpp
Comment (jameslamb) [inline src/io/config.cpp:243]: I was surprised to see this being removed, but I *think* it's ok. I couldn't find any cases of `Config.Set()` being called without `Config::Str2Map()` being called, and we have many tests on verbosity in the R and Python package which are passing.

Just wanted to say that out loud, for the benefit of other reviewers.

**Review by jameslamb — APPROVED** [review]
Changes look good to me, thanks! Since `Config` is so central to the codebase, I'd like one other maintainer to review and approve before we merge this.

File: src/io/config.cpp
Comment (StrikerRUS) [inline src/io/config.cpp:48]: If params have both `verbose` and `verbosity` keys I think we should still emit the following warning for this case
https://github.com/microsoft/LightGBM/blob/1cc9f9dcee15586aefdd8271a1c04ad010d3c53a/include/LightGBM/config.h#L1145-L1147
to inform user that `verbosity` has higher precedence over `verbose`.

Comment (StrikerRUS) [inline src/io/config.cpp:48]: I believe we should pop `verbose` and `verbosity` here to not process them twice later
https://github.com/microsoft/LightGBM/blob/1cc9f9dcee15586aefdd8271a1c04ad010d3c53a/src/io/config.cpp#L45

File: src/io/config.cpp
Comment (StrikerRUS) [inline src/io/config.cpp:46]: Shouldn't we update the following function as well?
https://github.com/microsoft/LightGBM/blob/1cc9f9dcee15586aefdd8271a1c04ad010d3c53a/src/application/application.cpp#L50

File: src/io/config.cpp
Comment (StrikerRUS) [inline src/io/config.cpp:43]: Looks like the following warnings in `KV2Map()` function are still cannot be suppressed by passing `verbosity` param because `// define verbosity and set logging level` happens later in the code
https://github.com/microsoft/LightGBM/blob/1cc9f9dcee15586aefdd8271a1c04ad010d3c53a/src/io/config.cpp#L28-L35

File: tests/python_package_test/test_engine.py
Comment (StrikerRUS) [inline tests/python_package_test/test_engine.py:3605]: I believe this will be more correct.
```suggestion
    if verbosity >= 0:
        assert expected_msg in stdout
    else:
        assert re.search(r'\[LightGBM\]', stdout) is None
```

File: tests/python_package_test/test_engine.py
Comment (StrikerRUS) [inline tests/python_package_test/test_engine.py:3586]: Since the original issue is quite "sensitive", could you please add a test that checks behavior when user passes multiple aliases for the `verbosity` param?
Refer to .

**Review by StrikerRUS — COMMENTED** [review]
Thanks for this PR!
I'm sorry, config initialization routine is still not clear enough for me. However, I left some comments that may help to spot uncovered by this PR cases.

File: src/io/config.cpp
Comment (jmoralez) [reply src/io/config.cpp:43]: These warnings come from duplicated parameters, not aliases. I can try to address them as well in this PR if its in the scope.

File: src/io/config.cpp
Comment (jmoralez) [reply src/io/config.cpp:46]: I believe if I move the logic into `KeyAliasTransform` this function will work as expected as well, except for the duplicate argument warnings.

File: src/io/config.cpp
Comment (jmoralez) [reply src/io/config.cpp:48]: I wanted to leave them because I'm not sure if they're used some place after but I can drop them here if this is the only place they're used.

File: src/io/config.cpp
Comment (StrikerRUS) [reply src/io/config.cpp:46]: I think it's good idea to move the logic into `KeyAliasTransform`.

File: src/io/config.cpp
Comment (StrikerRUS) [reply src/io/config.cpp:43]: I believe it's related.

**Comment by jmoralez** [comment]
I've changed the approach, now:
* Every parameter is saved to an `unordered_multimap` which can have several values for the same key, this is so we can see if there are any duplicate parameters after parsing them.
* Once we have all the parameters in the multimap we check for `verbose` and `verbosity` and set the logging level.
* We then move the multimap to a regular map and warn about duplicate arguments. This step takes a random one if two or more are provided (we can discuss this).
* Once the parameters are in a map we check for the aliases as before.
This of course uses more memory now but I couldn't think of a better way to handle the possible duplicate parameters. I'm open to adapt this to a better idea.

**Comment by jmoralez** [comment]
There are some issues right now with this because it prints the warnings several times and there seems to be a weird interaction with the dataset both in the python and R-packages where if you don't re-construct the dataset, verbosity -1 doesn't remove the alias warnings. I have to look into that.

**Comment by jmoralez** [comment]
Seems like the duplicate warnings about aliases is on master as well. I'll open an issue.

File: src/io/config.cpp
Comment (StrikerRUS) [reply src/io/config.cpp]: I agree with you. I believe each warning is helpful, but unfortunately LightGBM users want to completely silence their program (according to the issues they create in the repo).

**Review by guolinke — APPROVED** [review]
Thank you!

File: src/application/application.cpp
Comment (StrikerRUS) [inline src/application/application.cpp:73]: Ah, this warning will be emitted despite the verbosity setting.

**Comment by StrikerRUS** [comment]
Could you please resolve merge conflicts?

File: src/io/config.cpp
Comment (StrikerRUS) [inline src/io/config.cpp:32]: I believe it may be confusing that this function designed for the general purpose (according to its name) checks that value should be `int` typed.

**Review by StrikerRUS — COMMENTED** [review]
Thanks a lot for this PR!
I don't have any other additional comments except a one below about the function naming.

But I think I'm not qualified to formally approve these mostly cpp changes.
✔️

File: src/io/config.cpp
Comment (jmoralez) [reply src/io/config.cpp]: I modified the name to `GetFirstValue`. Given that the `out` param is an `int*` I think it's expected to check that the value is an `int`.

File: src/io/config.cpp
Comment (guolinke) [inline src/io/config.cpp:32]: maybe rename to "GetFirstIntValue" ? as it only supports int type?

**Review by guolinke — APPROVED** [review]
Looks good to me, except a minor naming

File: src/io/config.cpp
Comment (jmoralez) [reply src/io/config.cpp]: changed to `GetFirstValueAsInt` in e9a7a0f

**Comment by StrikerRUS** [comment]
Can we merge this PR?

**Comment by jameslamb** [comment]
> Seems like the duplicate warnings about aliases is on master as well. I'll open an issue.

did this happen? I couldn't find an issue like that from search and don't see one linked here.

> Can we merge this PR?

Based on what I see in the diff for unit tests, I think this can be merged. It does seem like the warnings described in https://github.com/microsoft/LightGBM/issues/3641#issuecomment-898092405 (which led to #4518) can now be suppressed with `verbosity`.

But I'll defer to @jmoralez . To be honest, there have been significant changes since I first approved and it's difficult for me to understand the current state of this PR a bit.

For example, the conversation in https://github.com/microsoft/LightGBM/pull/5253#discussion_r889728704 is unresolved.

**Comment by jmoralez** [comment]
> did this happen?

Yes, I opened https://github.com/microsoft/LightGBM/issues/5332.

The duplicate parameter warnings were also addressed in this PR. The `KeepFirstValues` function keeps the first value defined for each parameter and warns about the rest (if there are any). This allows to override parameters defined in a configuration file through arguments in the CLI, or if defined twice in the parameter list in the R-package, the first one takes precedence and a warning is printed for the second one.

My opinion is that this is ready to be merged, but of course I'm biased haha.

File: src/io/config.cpp
Comment (jmoralez) [reply src/io/config.cpp]: Addressed by having an unordered_map of vectors where we store all parameters for as many times as they're defined, set the verbosity taking the first values of `verbose` and `verbosity` and then de-duplicating with the `KeepFirstValues` function, where warning level logs are printed but after the verbosity has been set, so these can be disabled as well.

File: src/io/config.cpp
Comment (jmoralez) [reply src/io/config.cpp]: I left them so that we can warn about duplicates later.

Comment (jmoralez) [reply src/io/config.cpp]: This is done in the `KeyAliasTransform` function after setting the verbosity and de-duplicating.

**Review by jameslamb — CHANGES_REQUESTED** [review]
> Yes, I opened https://github.com/microsoft/LightGBM/issues/5332.

ah ok, thanks for pointing that out! I just got unlucky with the search terms I chose 😅 

If that issue had mentioned this PR with `#5253` or if a comment had been posted on this PR saying "created `#5332`", it would have been easier for me to find.

> The duplicate parameter warnings were also addressed in this PR

Sorry, I'm still confused...do you mean specifically that as of this PR being merged, we could also close #5332?

If so, please:

* add "fixes #5332" to the description
* add one more test confirming that the duplicate warnings are no longer raised

**Comment by jmoralez** [comment]
> do you mean specifically that as of this PR being merged, we could also close https://github.com/microsoft/LightGBM/issues/5332?

No, I meant that this was originally only going to remove warnings for parameter aliases, i.e. you set `bagging_fraction` and `subsample`, but in https://github.com/microsoft/LightGBM/pull/5253#discussion_r889728704 it was suggested that this also addressed warnings for duplicate parameters, i.e. you set `bagging_fraction=0.7` and `bagging_fraction=0.8`, so that conversation was resolved because it was also addressed.

The duplicate prints (#5332) is not solved yet.

**Review by jameslamb — APPROVED** [review]
Thanks for this and sorry for taking so long to come back and re-review! I understand now what his PR does and doesn't fix.

Let's merge this...it's only changing internal details, so easy to revise any of these design decisions in the future if we decide a different pattern is preferable for changing other logging-related stuff.

Awesome contribution, @jmoralez ! People have been asking for this for a long time 😬

**Pull request merged**
